File: macos/Classes/screen/ScreenProtector.swift
import Cocoa

/// Prevents screen capture of the application window on macOS.
///
/// Uses NSWindow's sharingType property to control screen capture behavior:
/// - `.none` — window content is excluded from screenshots and screen recording
/// - `.readOnly` — normal behavior (default)
///
/// macOS 10.14+: `NSWindow.SharingType.none` prevents the window from
/// appearing in screenshots, screen recordings, and screen sharing sessions.
/// The window appears as a black rectangle in captures.
class ScreenProtector {
    private var isEnabled = false

    /// Enable screen protection on all application windows.
    func enable() -> Bool {
        guard !isEnabled else { return true }

        DispatchQueue.main.async {
            for window in NSApplication.shared.windows {
                window.sharingType = .none
            }
        }

        isEnabled = true
        return true
    }

    /// Disable screen protection.
    func disable() -> Bool {
        guard isEnabled else { return true }

        DispatchQueue.main.async {
            for window in NSApplication.shared.windows {
                window.sharingType = .readOnly
            }
        }

        isEnabled = false
        return true
    }

    /// Whether screen protection is currently active.
    var isActive: Bool {
        return isEnabled
    }
}

File: windows/screen/screen_protector.cpp
#include "screen_protector.h"

#include <windows.h>

// WDA_EXCLUDEFROMCAPTURE is available from Windows 10 version 2004 (build 19041)
#ifndef WDA_EXCLUDEFROMCAPTURE
#define WDA_EXCLUDEFROMCAPTURE 0x00000011
#endif

namespace flutter_neo_shield {

ScreenProtector::ScreenProtector(flutter::PluginRegistrarWindows *registrar)
    : registrar_(registrar) {}

/// Enable screen capture protection.
///
/// WDA_EXCLUDEFROMCAPTURE (Windows 10 2004+): Window is completely excluded
/// from all capture APIs. Most secure option — window doesn't appear at all.
///
/// WDA_MONITOR (older Windows 10): Window appears as black in captures.
/// Less ideal but still prevents content leakage.
bool ScreenProtector::Enable() {
  if (is_active_) return true;

  HWND hwnd = GetFlutterWindow();
  if (!hwnd) return false;

  // Try WDA_EXCLUDEFROMCAPTURE first (Windows 10 2004+)
  if (::SetWindowDisplayAffinity(hwnd, WDA_EXCLUDEFROMCAPTURE)) {
    is_active_ = true;
    return true;
  }

  // Fall back to WDA_MONITOR (Windows 10 1709+)
  if (::SetWindowDisplayAffinity(hwnd, WDA_MONITOR)) {
    is_active_ = true;
    return true;
  }

  return false;
}

/// Disable screen capture protection.
bool ScreenProtector::Disable() {
  if (!is_active_) return true;

  HWND hwnd = GetFlutterWindow();
  if (!hwnd) return false;

  if (::SetWindowDisplayAffinity(hwnd, WDA_NONE)) {
    is_active_ = false;
    return true;
  }

  return false;
}

bool ScreenProtector::IsActive() const {
  return is_active_;
}

/// Get the Flutter application's main window handle.
HWND ScreenProtector::GetFlutterWindow() const {
  return registrar_->GetView()->GetNativeWindow();
}

}  // namespace flutter_neo_shield
